app/src/main/java/org/wikipedia/gallery/ImageInfo.kt

```diff
@@ -46,16 +46,25 @@ class ImageInfo {
     val width = 0
     val height = 0
 
-    val bestDerivative get() = derivatives.lastOrNull()
+    fun getBestDerivativeForSize(widthDp: Int): Derivative? {
+        var derivative: Derivative? = null
+        derivatives.forEach {
+            if (it.width > 0 && it.width < widthDp) {
+                if (derivative == null || it.width > derivative!!.width) {
+                    derivative = it
+                }
+            }
+        }
```

Comment:
Would it be better to use `filter` and `mayByOrNull`?
```
derivatives.filter { it.width < widthDp && it.width < widthDp }.maxByOrNull { it.width }
```

I tried a couple of times and it looks like using `filter` and `maxByOrNull` can mostly improve the efficiency
```
forEach                581.093us
filter+mayByOrNull     53.75us

forEach                62.032us
filter+mayByOrNull     67.604us

forEach                51.51us
filter+mayByOrNull     50.781us

forEach                118.489us
filter+mayByOrNull     57.239us
```

Reply:
Thanks for the information, that explains the benefit of using the original logic you put in the code! I checked the library and you are correct!